### PR TITLE
Add support for WithTags extension method

### LINKED ISSUE #35310: Add support for using endpoint metadata to add tags to an endpoint for use in OpenAPI descriptions
OpenAPI has a concept of [tags](https://swagger.io/docs/specification/grouping-operations-with-tags/) that allows for flexible grouping of APIs based on their associated tags.

Today, libraries like Swashbuckle use `ApiDescription.ActionDescriptor.RouteValues["controller"]` as the [default tag selector](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/cf7c50b70390a296f748d9a1f894823e2d9c7280/src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs#L80). This is the reason why minimal actions have their "controller" route value manually set to `IWebHostingEnvironment.ApplicationName` today when [they're populated in `ApiExplorer`](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L77).

We should enable users to be able to set tags for their endpoints via metadata and have those tags flow via `ApiExplorer` automatically to OpenAPI libraries like Swashbuckle, e.g.

``` c#
app.MapGet("/", () => "Hello World!")
   .WithTags("Examples");

app.MapGet("/todos/sample", () => new[] {
        new Todo { Id = 1, Title = "Do this" },
        new Todo { Id = 2, Title = "Do this too" }
    })
   .WithTags("Examples", "TodoApi");
```
## Proposed API Changes
``` diff
namespace Microsoft.AspNetCore.Http
{
  public static class OpenApiEndpointConventionBuilderExtensions
  {
+    public static MinimalActionEndpointConventionBuilder WithTags(this MinimalActionEndpointConventionBuilder builder, params string[] tags);
  }

+  [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method | AttributeTargets.Delegate)]
+  public class TagsAttribute : Attribute, ITagsMetadata
+  {
+      public TagsAttribute(params string[] tags);
+      public IList<string> Tags { get; }
+  }
  
+  public interface ITagsMetadata
+  {
+      IList<string> Tags { get; }
+  }
}
```

Issue to track corresponding Swashbuckle change: domaindrivendev/Swashbuckle.AspNetCore#2192

Example implementation can be found at https://github.com/DamianEdwards/MinimalApiPlayground:
- [`WithTags` extension method](https://github.com/DamianEdwards/MinimalApiPlayground/blob/main/src/MinimalApiPlayground/Properties/EndpointConventionBuilderExtensions.cs#L8)
- [TagsAttribute](https://github.com/DamianEdwards/MinimalApiPlayground/blob/main/src/MinimalApiPlayground/Properties/TagsAttribute.cs#L5)
- [ITagsMetadata](https://github.com/DamianEdwards/MinimalApiPlayground/blob/main/src/MinimalApiPlayground/Properties/TagsAttribute.cs#L15)
- [Swashbuckle tags selector](https://github.com/DamianEdwards/MinimalApiPlayground/blob/main/src/MinimalApiPlayground/Properties/OpenApi.cs#L82)

image

### PR DESCRIPTION
**Pull request by captainsafia**
## Description

This PR adds support for a new `WithTags` extension method that can be used to specify the tag groupings associated with a 

```csharp
app.MapGet("/", () => "Hello World!")
   .WithTags("Examples");

app.MapGet("/todos/sample", () => new[] {
        new Todo { Id = 1, Title = "Do this" },
        new Todo { Id = 2, Title = "Do this too" }
    })
   .WithTags("Examples", "TodoApi");
```

For API review, see https://github.com/dotnet/aspnetcore/issues/35693.

## Customer Impact

This PR makes it easier for developers building APIs with minimal APIs and documenting them with the OpenAPI spec to group correlated APIs into collections. Currently, the only option available uses the controller name to set the API group in the OpenAPI spec which is not applicable to Minimal APIs which leaves developers with no other options.

## Regression?
- [ ] Yes
- [X] No

## Risk
- [ ] High
- [ ] Medium
- [X] Low

Low risk because:
- New APIs don't have a high surface area for bugs to occur in.

## Verification
- [X] Manual (required)
- [X] Automated

## Packaging changes reviewed?
- [ ] Yes
- [ ] No
- [X] N/A

Addresses https://github.com/dotnet/aspnetcore/issues/35310

### CODE REVIEW

File: src/Http/Http/src/TagsAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        public TagsAttribute(params string[] tags)
+        {
+            Tags = new List<string>(tags);
+        }
+
+        /// <summary>
+        /// Gets the collection of tags associated with the endpoint.
+        /// </summary>
+        public IReadOnlyList<string> Tags { get; }
+    }
```
Comment (Kahbazi) [inline src/Http/Http/src/TagsAttribute.cs:27]: Helps https://github.com/dotnet/aspnetcore/issues/34604 in order to discover the endpoint metadata.
```suggestion
    }
    
    /// <inheritdoc />
    public override string ToString()
    {
        return $"Tags: {string.Join(",", Tags)}";
    }
```

File: src/Http/Http.Extensions/src/TagsAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    /// <summary>
+    /// Specifies a collection of tags in <see cref="Endpoint.Metadata"/>.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Delegate | AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+    public sealed class TagsAttribute : Attribute, ITagsMetadata
+    {
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="tags"></param>
```
Comment (halter73) [inline src/Http/Http.Extensions/src/TagsAttribute.cs:18]:
```suggestion
        /// <summary>
        /// Initializes an instance of the <see cref="TagsAttribute"/>.
        /// </summary>
        /// <param name="tags">The tags associated with the endpoint.</param>
```

Feel free to improve on this.

File: src/Http/Http/src/TagsAttribute.cs (hunk 1 above)
Comment (captainsafia) [reply src/Http/Http/src/TagsAttribute.cs]: Thanks for the recommendation! I think it'll be easier if we batch all the changes related to #34604 above into the same PR and review collectively.

Also, I suspect that maybe we want to use a `DebuggerDisplay` string for this as well.

File: src/Http/Routing/src/Builder/OpenApiDelegateEndpointConventionBuilderExtensions.cs [hunk 3]
```diff
@@ -8,13 +8,19 @@
...
-    /// Extension methods for adding response type metadata to endpoints.
+    /// Extension methods for adding <see cref="Endpoint.Metadata"/> that is
+    /// meant to be consumed OpenAPI libraries.
     /// </summary>
     public static class OpenApiDelegateEndpointConventionBuilderExtensions
     {
         /// <summary>
         /// Adds the <see cref="ITagsMetadata"/> to <see cref="EndpointBuilder.Metadata"/> for all builders
         /// produced by <paramref name="builder"/>.
+        /// <remarks>
```
Comment (DamianEdwards) [inline src/Http/Routing/src/Builder/OpenApiDelegateEndpointConventionBuilderExtensions.cs:19]: These shouldn't be nested but rather a sibling of the `<summary>` element

File: src/Http/Http.Extensions/src/TagsAttribute.cs [hunk 4]
```diff
@@ -8,14 +8,19 @@ namespace Microsoft.AspNetCore.Http
 {
     /// <summary>
     /// Specifies a collection of tags in <see cref="Endpoint.Metadata"/>.
+    /// <remarks>
```
Comment (DamianEdwards) [inline src/Http/Http.Extensions/src/TagsAttribute.cs:11]: These shouldn't be nested but rather a sibling of the `<summary>` element

**Comment by TanayParikh** [comment]
Due to the public api unshipped changes, this'll have to be force merged (upon approval), like we did for: https://github.com/dotnet/aspnetcore/pull/35979. Feel free to ping when it's ready 😄

**Review by BrennanConroy — APPROVED** [review]

**Comment by TanayParikh** [comment]
Only API checks are failing:
image

Force merging by request as this API has been cleared for 6.0.

**Pull request merged**